## Assignement 1

In [2]:
import nltk
from nltk.corpus import webtext
nltk.download('webtext')

[nltk_data] Downloading package webtext to
[nltk_data]     /home/frederico/nltk_data...
[nltk_data]   Package webtext is already up-to-date!


True

In [3]:
forum = webtext.raw('firefox.txt')
print(forum[:100])

Cookie Manager: "Don't allow sites that set removed cookies to set future cookies" should stay check


In [18]:
import re

http_urls = re.findall('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', forum)
www_urls = re.findall('www\.(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', forum)
#regex = '(?:https?:\/\/)?(?:www\.)?[-a-zA-Z0-9@:%._\+~#=]{1,256}\.[a-zA-Z0-9()]{1,6}'
#urls = re.findall(regex, forum)
urls = http_urls + www_urls
shortcuts = re.findall('(?:ctrl|shift|alt)+\s*\+\s*[a-z+\-.,/]', forum, flags=re.IGNORECASE)
print(urls)
print(shortcuts)

['http://www.scripting.com/misc/msswitchad', 'http://www.watch.impress.co.jp', 'http://bugzilla.mozilla.org', 'http://www.http://mozilla.org', 'http://www.peterre.com', 'http://texturizer.net/firebird', 'http://foo', 'http://http://', 'http://james', 'http://www.lexis.com', 'http://www.woolworth.de', 'http://ftp.mozilla.org/pub/mozilla.org/firebird/nightly', 'http://http://', 'http://extensionroom.mozdev.org/more-info', 'http://www.odeon.co.uk/odeon', 'http://www.cctvusa.com', 'http://www.trenitalia.com/home/it', 'https://www.fortify.net', 'http://irc-galleria.net', 'http://http', 'http://www.mozilla.org/products', 'http://labs.google.com/cgi-bin', 'http://www.timbressuisses.ch', 'https://www.eposasp.com/ebpp', 'www.scripting.com/misc/msswitchad', 'www.foo.com', 'www.localhost.net.au', 'www.watch.impress.co.jp', 'www.*.com', 'www.aol.com', 'www.php.net', 'www.fnac.fr', 'www.http://mozilla.org', 'www.hvv.de', 'www.petetownshend.co.uk', 'www.google.com', 'www.wamu.com)', 'www.excite.com'

## Assignement 2

In [38]:
import nltk
from nltk.corpus import gutenberg
nltk.download('gutenberg')


[nltk_data] Downloading package gutenberg to
[nltk_data]     /home/frederico/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!


True

In [39]:
hamlet = gutenberg.raw('shakespeare-hamlet.txt')
print(hamlet[:100])

[The Tragedie of Hamlet by William Shakespeare 1599]


Actus Primus. Scoena Prima.

Enter Barnardo a


#### 3 classes tagger

In [82]:
import nltk
from nltk.tag.stanford import StanfordNERTagger

PATH_TO_JAR = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/stanford-ner.jar'
PATH_TO_MODEL = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/classifiers/english.all.3class.distsim.crf.ser.gz'
NER = StanfordNERTagger(model_filename=PATH_TO_MODEL,path_to_jar=PATH_TO_JAR, encoding='utf-8')


In [88]:
words = nltk.wordpunct_tokenize(hamlet)
tagged = NER.tag(words)
people = []

for (word,label) in tagged:
    if label == 'PERSON':
        people.append(word)
people = list(set(people))
print(people[:10])
print(len(people))

['Voltemand', 'Horatio', 'Sallets', 'Carbuncles', 'Controuersie', 'Vnckle', 'Scullion', 'Fox', 'Throate', 'Pesant']
498


#### 4 classes tagger

In [87]:
PATH_TO_JAR = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/stanford-ner.jar'
PATH_TO_MODEL = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/classifiers/english.conll.4class.distsim.crf.ser.gz'
NER = StanfordNERTagger(model_filename=PATH_TO_MODEL,path_to_jar=PATH_TO_JAR, encoding='utf-8')

tagged = NER.tag(words)
people = []

for (word,label) in tagged:
    if label == 'PERSON':
        people.append(word)
people = list(set(people))
print(people[:10])
print(len(people))

['Voltemand', 'Horatio', 'Sallets', 'Carbuncles', 'Controuersie', 'Vnckle', 'Scullion', 'Fox', 'Throate', 'Pesant']
498


#### 7 classes tagger

In [86]:
PATH_TO_JAR = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/stanford-ner.jar'
PATH_TO_MODEL = '/home/frederico/Desktop/DS/stanford-ner-2020-11-17/classifiers/english.muc.7class.distsim.crf.ser.gz'
NER = StanfordNERTagger(model_filename=PATH_TO_MODEL,path_to_jar=PATH_TO_JAR, encoding='utf-8')

tagged = NER.tag(words)
people = []

for (word,label) in tagged:
    if label == 'PERSON':
        people.append(word)
people = list(set(people))
print(people[:10])
print(len(people))

['leaue', 'Sonne', 'Horatio', 'Keepes', 'Barnardo', 'Gertrude', 'Businesse', 'Scullion', 'Leuies', 'Scul']
226


## Assignement 3

In [1]:
import nltk
from nltk.corpus import reuters
nltk.download('reuters')

[nltk_data] Downloading package reuters to
[nltk_data]     /home/frederico/nltk_data...
[nltk_data]   Package reuters is already up-to-date!


True

In [2]:
rts = reuters.raw('test/14826')
print(rts[:100])

ASIAN EXPORTERS FEAR DAMAGE FROM U.S.-JAPAN RIFT
  Mounting trade friction between the
  U.S. And Ja


In [3]:
print(reuters.categories())

['acq', 'alum', 'barley', 'bop', 'carcass', 'castor-oil', 'cocoa', 'coconut', 'coconut-oil', 'coffee', 'copper', 'copra-cake', 'corn', 'cotton', 'cotton-oil', 'cpi', 'cpu', 'crude', 'dfl', 'dlr', 'dmk', 'earn', 'fuel', 'gas', 'gnp', 'gold', 'grain', 'groundnut', 'groundnut-oil', 'heat', 'hog', 'housing', 'income', 'instal-debt', 'interest', 'ipi', 'iron-steel', 'jet', 'jobs', 'l-cattle', 'lead', 'lei', 'lin-oil', 'livestock', 'lumber', 'meal-feed', 'money-fx', 'money-supply', 'naphtha', 'nat-gas', 'nickel', 'nkr', 'nzdlr', 'oat', 'oilseed', 'orange', 'palladium', 'palm-oil', 'palmkernel', 'pet-chem', 'platinum', 'potato', 'propane', 'rand', 'rape-oil', 'rapeseed', 'reserves', 'retail', 'rice', 'rubber', 'rye', 'ship', 'silver', 'sorghum', 'soy-meal', 'soy-oil', 'soybean', 'strategic-metal', 'sugar', 'sun-meal', 'sun-oil', 'sunseed', 'tea', 'tin', 'trade', 'veg-oil', 'wheat', 'wpi', 'yen', 'zinc']


In [4]:
documents = [(list(reuters.words(fileid)), category)
             for category in reuters.categories()
             for fileid in reuters.fileids(category)]

# first 10 words of 1 document
print(documents[0][0][:10])
# category of 1 document
print(documents[0][1])

import random
random.shuffle(documents)

['SUMITOMO', 'BANK', 'AIMS', 'AT', 'QUICK', 'RECOVERY', 'FROM', 'MERGER', 'Sumitomo', 'Bank']
acq


#### Most common words among all words

In [5]:
all_words = nltk.FreqDist(w.lower() for w in reuters.words())
# 5 most common words
print(all_words.most_common(5))
word_features = list(all_words)[:2000]

# check if the document has the most common words among it's words
def document_features(document):
    document_words = set(document) 
    features = {}
    for word in word_features:
        features['contains({})'.format(word)] = (word in document_words)
    return features


[('.', 94687), (',', 72360), ('the', 69277), ('of', 36779), ('to', 36400)]


#### First extractor

In [20]:
featuresets = [(document_features(d), c) for (d,c) in documents]
print(len(featuresets))
train_set, test_set = featuresets[5000:], featuresets[:200]
classifier = nltk.NaiveBayesClassifier.train(train_set)

In [9]:
# example
print(classifier.classify(document_features(rts)))
print(reuters.categories('test/14826'))

classifier.show_most_informative_features(5)

gas
['trade']
Most Informative Features
          contains(palm) = True           palm-o : earn   =   1614.4 : 1.0
        contains(rubber) = True           rubber : earn   =   1433.8 : 1.0
          contains(zinc) = True             zinc : earn   =   1396.7 : 1.0
      contains(supplies) = True           propan : earn   =   1372.8 : 1.0
        contains(coffee) = True           coffee : earn   =   1355.1 : 1.0


In [10]:
print(nltk.classify.accuracy(classifier, test_set))

0.54


#### Second extractor

#### Most common words among the document

In [18]:
def alternative_document_features(document):
    document_words = nltk.FreqDist(w.lower() for w in document)
    word_features = list(document_words)[:2000]
    return dict([(word, True) for word in word_features])


In [19]:
featuresets = [(alternative_document_features(d), c) for (d,c) in documents]
print(len(featuresets))
print(featuresets[0])
train_set, test_set = featuresets[5000:], featuresets[:200]
classifier = nltk.NaiveBayesClassifier.train(train_set)

13328
({'the': True, ',': True, 'stock': True, '.': True, '-': True, 'dividend': True, 'on': True, 'hydraulic': True, 'said': True, 'split': True, 'of': True, 'will': True, 'april': True, 'share': True, '3': True, 'its': True, 'a': True, 'common': True, 'quarterly': True, 'cash': True, 'to': True, 'be': True, 'cts': True, 'per': True, 'for': True, 'it': True, '50': True, "'": True, 's': True, 'payable': True, 'stockholders': True, 'record': True, 'that': True, 'outstanding': True, 'company': True, '&': True, 'lt': True, ';': True, 'thc': True, '>': True, 'splits': True, '2': True, 'hikes': True, 'co': True, 'board': True, 'approved': True, 'three': True, 'two': True, 'and': True, 'increased': True, 'occur': True, 'through': True, 'pct': True, 'distribution': True, '30': True, '15': True, 'is': True, 'paid': True, 'pre': True, 'shares': True, 'are': True, 'currently': True, '54': True, '75': True, 'up': True, 'from': True, '52': True, 'represent': True, '36': True, 'after': True}, 'earn

In [20]:
# example
print(classifier.classify(alternative_document_features(rts)))
print(reuters.categories('test/14826'))

classifier.show_most_informative_features(5)

dfl
['trade']
Most Informative Features
                  coffee = True           coffee : earn   =   1622.3 : 1.0
                 propane = True           propan : earn   =   1468.8 : 1.0
                    oats = True              oat : earn   =   1428.0 : 1.0
               argentine = True           lin-oi : earn   =   1360.0 : 1.0
                minister = True            nzdlr : earn   =   1360.0 : 1.0


In [21]:
print(nltk.classify.accuracy(classifier, test_set))

0.005


#### Discussion of results

Given the difference in results between both extractors, we can conclude that the first method of extracting information, that is, extracting the most common words across all texts and then check which of them are present for a given file, has a very supperior accuracy to the normal bag of words approach (54% accuracy vs 0.5% accuracy).
One possible reason for this might be that having a basis of possible words (the most common across all files) allows for a easier training of the model and for an easier recognition of the class of the file since we can focus on a smaller and more focused set of words.